org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java

```diff
@@ -0,0 +1,655 @@
+          }
+        }
+      }
+      Object[] resources = root.getNonJavaResources();
+
+      for (Object resource : resources) {
+        if (resource instanceof JarEntryFile) {
+          JarEntryFile file = (JarEntryFile) resource;
+          if (file.getFullPath().toOSString().equals(path)) {
+            return readFileContent(file);
+          }
+        }
+        if (resource instanceof JarEntryDirectory) {
+          JarEntryDirectory directory = (JarEntryDirectory) resource;
+          JarEntryFile file = findJarFile(directory, path);
+          if (file != null) {
+            return readFileContent(file);
+          }
+        }
+      }
+    } else {
+      return getContent(project, path, pm);
+    }
+    return null;
+  }
+
+  /**
+   * Computes content of java file by fqn.
+   *
+   * @param project project
+   * @param path path of the node
+   * @param pm progress monitor
+   * @return content of the library's node
+   * @throws JavaModelException if an exception occurs while accessing its corresponding resource
+   */
+  private static String getContent(IJavaProject project, String path, IProgressMonitor pm)
+      throws JavaModelException {
+    IType type = project.findType(path);
+    if (type == null) {
+      throw new JavaModelException(new JavaModelStatus(0, "Can't find type: " + path));
+    }
+
+    Location location = JDTUtils.toLocation(type);
+
+    ContentProviderManager handler = JavaLanguageServerPlugin.getContentProviderManager();
+    return handler.getContent(JDTUtils.toURI(location.getUri()), pm);
+  }
+
+  private static Object[] getPackageContent(IPackageFragment fragment, IProgressMonitor pm)
+      throws JavaModelException {
+    // hierarchical package mode
+    ArrayList<Object> result = new ArrayList<>();
+
+    getHierarchicalPackageChildren(
+        (IPackageFragmentRoot) fragment.getParent(), fragment, result, pm);
+    IClassFile[] classFiles = fragment.getClassFiles();
+    List<IClassFile> filtered = new ArrayList<>();
+    // filter inner classes
+    for (IClassFile classFile : classFiles) {
+      if (!classFile.getElementName().contains("$")) {
+        filtered.add(classFile);
+      }
+    }
+    Object[] nonPackages = concatenate(filtered.toArray(), fragment.getNonJavaResources());
+    if (result.isEmpty()) {
+      return nonPackages;
+    }
+    Collections.addAll(result, nonPackages);
+    return result.toArray();
+  }
+
+  private static JarEntryFile findJarFile(JarEntryDirectory directory, String path) {
+    for (IJarEntryResource children : directory.getChildren()) {
+      if (children.isFile() && children.getFullPath().toOSString().equals(path)) {
+        return (JarEntryFile) children;
+      }
+      if (!children.isFile()) {
+        JarEntryFile file = findJarFile((JarEntryDirectory) children, path);
+        if (file != null) {
+          return file;
+        }
+      }
+    }
+    return null;
+  }
+
+  private static IPackageFragmentRoot getPackageFragmentRoot(
```

Comment:
If the id is the handle identifier you can simply do: root= (IPackageFragmentRoot)JavaCore.create(id). No need to iterate over all roots.
In this way, we won't need the project to get the package fragment root for a library id. -> remove it from the parameters where applicable.

Reply:
Cool, thanks! Changed